Allow to ignore errors while pushing to MATERILIZED VIEW (and ignore for them for logs flushing)

azat opened the pull request:
Note, this had been implemented in a most simplest way, just to verify the idea.

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Allow to ignore errors while pushing to MATERILIZED VIEW (add new setting `materialized_views_ignore_errors`, by default to `false`, but it is set to `true` for flushing logs to `system.*_log` tables unconditionally)

This can be useful in the following scenarious:

- you want to duplicate the data to another table and you don't care
  about the errors
- you want to duplicate system.*_log to another server, you are adding
  materialized view that will push to Distributed table, but you don't
  want to miss original blocks in the local system.*_log
- you want to push some data to a 3d party service, using i.e. URL
  engine.

Fixes: #18407
Fixes: #21721
Fixes: #35837
Cc: @alexey-milovidov

azat committed:
Allow to ignore errors while pushing to MATERILIZED VIEW

This can be useful in the following scenarious:

- you want to duplicate the data to another table and you don't care
  about the errors
- you want to duplicate system.*_log to another server, you are adding
  materialized view that will push to Distributed table, but you don't
  want to miss original blocks in the local system.*_log
- you want to push some data to a 3d party service, using i.e. URL
  engine.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Set materialized_views_ignore_errors unconditionally for system.*_log tables

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

Algunenano commented on src/Processors/Transforms/buildPushingToViewsChain.cpp:
```diff
@@ -788,6 +791,13 @@ void FinalizingViewsTransform::work()
         auto & status = statuses[i];
         ++i;
 
+        if (status.exception && materialized_views_ignore_errors)
```
I'd expect to still see the failure in query_views_log, which I think this is skipping. Maybe merging the 2 ifs would be more clear?

Algunenano commented on docs/en/sql-reference/statements/create/view.md:
```diff
@@ -66,6 +66,10 @@ A materialized view is implemented as follows: when inserting data to the table
 
 Materialized views in ClickHouse are implemented more like insert triggers. If there’s some aggregation in the view query, it’s applied only to the batch of freshly inserted data. Any changes to existing data of source table (like update, delete, drop partition, etc.) does not change the materialized view.
+
+By default if the will be an error during pushing to the materialized view the data will not be inserted into the table to which the materialized view is attached. You can change this by setting `materialized_views_ignore_errors=true` setting for your `INSERT` query.
```
This description is unclear to me. Currently, if there is an error on a MV data will be ingested on the landing table (or at least some blocks), other MVs and it might or might not have ingested data on the MV target. The setting seems to be a "ignore any error" and push wherever it's possible.

azat replied:
Yeah, without transaction the behavior is undetermined, but since documentation has nothing about "transactions", then I will just write about non deterministic behavior by default and how it can be changed.

@Algunenano Does new documentation paragraphs more clear to you?

Algunenano replied:
Yes. Much better. Thanks

azat committed:
Add an entry to system.query_views_log regardless of materialized_views_ignore_errors

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Improve 02572_system_logs_materialized_views_ignore_errors test

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Improve documentation for materialized_views_ignore_errors

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Do not run 02572_system_logs_materialized_views_ignore_errors for Replicated

CI: https://s3.amazonaws.com/clickhouse-test-reports/46658/3a58951f6446b3aeca13cdc671322f459c4f77a7/stateless_tests__release__databasereplicated__[1/4].html
Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

Algunenano commented on src/Processors/Transforms/buildPushingToViewsChain.cpp:
```diff
@@ -684,10 +686,11 @@ void PushingToWindowViewSink::consume(Chunk chunk)
     : IProcessor(initPorts(std::move(headers)), {Block()})
     , output(outputs.front())
     , views_data(std::move(data))
+    , materialized_views_ignore_errors(materialized_views_ignore_errors_)
```
I think we can get away with not adding a new parameter to the class and instead read it from data:

```c++
bool materialized_views_ignore_errors = views_data->context->getSettingsRef().materialized_views_ignore_errors;
```

The less parameters the better IMO

azat replied:
Yes, make sense. Thanks.

azat committed:
Remove materialized_views_ignore_errors from FinalizingViewsTransform

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
>PullRequestCI / PerformanceComparisonX86-0 (pull_request) Failing after 67m

- #46641 

>Integration tests (asan) [6/6] — fail: 1, passed: 418, flaky: 0

- `test_postgresql_replica_database_engine_1/test.py::test_abrupt_connection_loss_while_heavy_replication` is flaky according to https://aretestsgreenyet.com/

>[Integration tests (tsan) [4/6] — fail: 1, passed: 257, flaky: 0](https://s3.amazonaws.com/clickhouse-test-reports/46658/d0b50adffa98b6db67586098430d8f08f970e174/integration_tests__tsan__[4/6].html)

```
E       AssertionError: assert {'snapshot_150.bin.zstd', '.snapshot_200.bin.zstd_LOCK', 'snapshot_100.bin.zstd', 'snapshot_50.bin.zstd'} == {'snapshot_150.bin.zstd', 'snapshot_100.bin.zstd', 'snapshot_200.bin.zstd', 'snapshot_50.bin.zstd'}
E         Extra items in the left set:
E         '.snapshot_200.bin.zstd_LOCK'
E         Extra items in the right set:
E         'snapshot_200.bin.zstd'
E         Full diff:
E           {
E         +  '.snapshot_200.bin.zstd_LOCK',
E            'snapshot_100.bin.zstd',
E            'snapshot_150.bin.zstd',
E         -  'snapshot_200.bin.zstd',
E            'snapshot_50.bin.zstd',
E           }
```

Also flaky, here is another [example](https://s3.amazonaws.com/clickhouse-test-reports/46287/977afe10093e94bf8edd2c2dea3f5ef0ba539ee9/integration_tests__tsan__[4/6].html)

>[Stress test (debug) — Killed by signal (in clickhouse-server.log)](https://s3.amazonaws.com/clickhouse-test-reports/46658/d0b50adffa98b6db67586098430d8f08f970e174/stress_test__debug_.html)

```
 [ 7103 ] {} <Fatal> BaseDaemon: ########################################
 [ 7103 ] {} <Fatal> BaseDaemon: (version 23.2.1.1, build id: 33EEA9875A313E568FE3D479B944CF7D43502EE7) (from thread 2519) (query_id: 000bfb36-4cf7-4e28-8451-5f94d59a2d62) (query: INSERT INTO test_1.in_02232 (key) VALUES) Received signal Illegal instruction (4)
 [ 7103 ] {} <Fatal> BaseDaemon: Illegal operand.
 [ 7103 ] {} <Fatal> BaseDaemon: Stack trace: 0x30ed51e8 0x30ed5704 0x30ed4e9b 0x21ad1205 0x1919229d 0x1919221d 0x191921b8 0x28687fca 0x28687f65 0x2868734c 0x34c97e83 0x1919d2d6 0x1d379e4c 0x21deb85d 0x21deb7ed 0x21deb795 0x21deb779 0x21deb56e 0x1919d314 0x1919d2b9 0x21dc6c2c 0x21dedcc0 0x21dedc8d 0x21dedc35 0x21dedc19 0x21deda0e 0x1919d314 0x1919d2b9 0x28e20a2c 0x28e24e0d 0x28e24db5 0x28e24d19 0x
 [ 7103 ] {} <Fatal> BaseDaemon: 4. /build/build_docker/../contrib/llvm-project/compiler-rt/lib/gwp_asan/platform_specific/utilities_posix.cpp:27: gwp_asan::die(char const*) @ 0x30ed51e8 in /usr/bin/clickhouse
 [ 7103 ] {} <Fatal> BaseDaemon: 5. /build/build_docker/../contrib/llvm-project/compiler-rt/lib/gwp_asan/platform_specific/guarded_pool_allocator_posix.cpp:98: gwp_asan::GuardedPoolAllocator::deallocateInGuardedPool(void*, unsigned long) const @ 0x30ed5704 in /usr/bin/clickhouse
 [ 7103 ] {} <Fatal> BaseDaemon: 6. /build/build_docker/../contrib/llvm-project/compiler-rt/lib/gwp_asan/guarded_pool_allocator.cpp:322: gwp_asan::GuardedPoolAllocator::deallocate(void*) @ 0x30ed4e9b in /usr/bin/clickhouse
 [ 7103 ] {} <Fatal> BaseDaemon: 7.1. inlined from /build/build_docker/../src/Common/memory.h:116: void Memory::deleteSized<>(void*, unsigned long)
```

>[Stress test (ubsan) — Killed by signal (in clickhouse-server.log)](https://s3.amazonaws.com/clickhouse-test-reports/46658/d0b50adffa98b6db67586098430d8f08f970e174/stress_test__ubsan_.html)

```
 [ 2779 ] {} <Fatal> : Logical error: \'Cache usage is not allowed (query_id: 95b34dcc-4883-462c-998c-76c2bc8b57b8)\'.
 [ 2697 ] {} <Fatal> : Logical error: \'Cache usage is not allowed (query_id: 95b34dcc-4883-462c-998c-76c2bc8b57b8)\'.
```

@kssenii Looks like #46521 was not enough

azat commented:
Test failures does not looks related (see details above)

azat commented:
Can someone take a look please?

kssenii committed:
Update docs/en/sql-reference/statements/create/view.md

azat commented on docs/en/sql-reference/statements/create/view.md:
```diff
@@ -66,6 +66,12 @@ A materialized view is implemented as follows: when inserting data to the table
 
 Materialized views in ClickHouse are implemented more like insert triggers. If there’s some aggregation in the view query, it’s applied only to the batch of freshly inserted data. Any changes to existing data of source table (like update, delete, drop partition, etc.) does not change the materialized view.
+
+Materialized views in ClickHouse does not have deterministic behaviour in case of errors. This means that blocks that had been already written will be preserved in the destination table, but all blocks after error will not.
```
Interesting, AFAICS more common usage is `does`, am I missing something?

```
azat:.../ch/clickhouse (master) [1]$ find docs -type f | xargs cat | grep -c -i 'clickhouse does '
30
azat:.../ch/clickhouse (master) [1]$ find docs -type f | xargs cat | grep -c -i 'clickhouse do '
0
```

kssenii replied:
But here `do` refers not to "ClickHouse" but to "views"

azat replied:
Oh, I see, indeed, make total sense, thanks.

azat committed:
Merge branch 'master' into mv-ignore-errors

azat commented:
Test failures does not looks related

>Stress test (ubsan) — Hung check failed, possible deadlock found (see hung_check.log)

```
Using queries from '/usr/share/clickhouse-test/queries' directory
Connecting to ClickHouse server...Connection error, will retry:  [Errno 104] Connection reset by peer
```

Server terminated correctly:

```
2023.03.09 10:23:55.867724 [ 1705 ] {} <Information> Application: Closed connections. But 13 remain. Tip: To increase wait time add to config: <shutdown_wait_unfinished>60</shutdown_wait_unfinished>
2023.03.09 10:23:55.867747 [ 1705 ] {} <Information> Application: Will shutdown forcefully.
2023.03.09 10:23:57.798581 [ 1704 ] {} <Information> Application: Child process exited normally with code 0.
```

Also the server was under memory pressure:

```
2023.03.09 10:23:55.620232 [ 1979 ] {} <Error> void DB::MergeTreeBackgroundExecutor<DB::DynamicRuntimeQueue>::routine(DB::TaskRuntimeDataPtr) [Queue = DB::DynamicRuntimeQueue]: Code: 241. DB::Exception: Memory limit (total) exceeded: would use 34.36 GiB (attempt to allocate chunk of 1532111 bytes), maximum: 34.29 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or overcommit denominator are set to zero. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
```

>Upgrade check (debug) — Error message in clickhouse-server.log (see upgrade_error_messages.txt)

```
2023.03.09 04:36:16.081617 [ 294814 ] {} <Error> InterserverIOHTTPHandler: Code: 232. DB::Exception: No part all_44_50_1 in table. (NO_SUCH_DATA_PART), Stack trace (when copying this message, always include the lines below): 2023.03.09 04:36:16.348724 [ 295093 ] {} <Error> bd58ecd2-769e-42ef-96b9-a572e92a6d58::all_44_50_1 (MergeFromLogEntryTask): virtual bool DB::ReplicatedMergeMutateTaskBase::executeStep(): Code: 86. DB::HTTPException: Received error from remote server /?endpoint=DataPartsExchange%3A%2Ftest%2F01921_con 2023.03.09 04:36:25.627652 [ 294902 ] {} <Error> DiskObjectStorageTransaction: An error occurred while executing transaction\'s operation #0 (RemoveRecursiveObjectStorageOperation (path: store/362/36249891-4824-4a9c-96ba-faf7416e1583/tmp-fetch_all_65_65_0/)): Code: 107. DB::ErrnoException: Cannot ope 2023.03.09 04:36:26.069342 [ 294902 ] {} <Error> test_9grhzavs.ttl_table3 (ReplicatedMergeTreeCleanupThread): void DB::ReplicatedMergeTreeCleanupThread::run(): Code: 107. DB::ErrnoException: Cannot open file /var/lib/clickhouse/disks/s3/store/362/36249891-4824-4a9c-96ba-faf7416e1583/tmp-fetch_all_65 2023.03.09 04:36:26.076173 [ 294964 ] {} <Error> DiskObjectStorageTransaction: An error occurred while executing transaction\'s operation #0 (RemoveRecursiveObjectStorageOperation (path: store/536/536a197a-b5de-4bfc-9c71-1db5cbe7ce3b/tmp-fetch_all_55_55_0/)): Code: 107. DB::ErrnoException: Cannot ope 2023.03.09 04:36:26.078578 [ 294964 ] {} <Error> test_ojl3s5p9.ttl_table1 (ReplicatedMergeTreeCleanupThread): void DB::ReplicatedMergeTreeCleanupThread::run(): Code: 107. DB::ErrnoException: Cannot open file /var/lib/clickhouse/disks/s3/store/536/536a197a-b5de-4bfc-9c71-1db5cbe7ce3b/tmp-fetch_all_55
```

- #47315
